addons/binding/org.openhab.binding.homie/src/main/java/org/openhab/binding/homie/handler/HomieDeviceHandler.java

```diff
@@ -0,0 +1,421 @@
+            }
+        }
+
+        protected final void handleCommand(Command command) {
+            if (command == RefreshType.REFRESH) {
+                try {
+                    connection.listenForNode(getThing().getUID().getId(), nodeId, this);
+                } catch (MqttException e) {
+                    logger.error("Error handling command", e);
+                }
+            } else {
+                onHandleCommand(command);
+            }
+        }
+
+        protected abstract void onHandleCommand(Command command);
+
+        public String getNodeId() {
+            return nodeId;
+        }
+
+        public String getCategory() {
+            return channelCategory;
+        }
+
+        public ChannelUID getChannelUID() {
+            return channelUID;
+        }
+
+        public Channel getChannel() {
+            return getThing().getChannel(channelId);
+        }
+
+        @Override
+        public void messageArrived(String topic, MqttMessage mqttMessage) throws Exception {
+            String msg = mqttMessage.toString();
+            HomieTopic ht = topicParser.parse(topic);
+            onMqttMessage(ht, msg);
+        }
+
+        protected abstract void onMqttMessage(HomieTopic topic, String message);
+
+        public void setProperties(NodePropertiesList properties) {
+            this.properties = properties;
+        }
+
+    }
+
+    private class ESHNodeHandler extends NodeHandler {
+
+        private String unit;
+        private BigDecimal min;
+        private BigDecimal max;
+        private BigDecimal step;
+        private String itemType = "String";
+        private boolean isReadonly;
+        private String description = "";
+
+        public ESHNodeHandler(String nodeId, HomieConfiguration config, String category) throws MqttException {
+            super(nodeId, config, category);
+            description = "Homie Node '" + getNodeId() + "'";
+        }
+
+        @Override
+        public synchronized void init() throws MqttException {
+            isReadonly = getProperties().isPropertySettable(ESH_VALUE_TOPIC);
+            super.init();
+
+        }
+
+        private void updateChannelType() {
+            ChannelTypeUID channelTypeUID = typeProvider.createChannelTypeBySettings(unit, min, max, step, itemType,
+                    isReadonly, getCategory());
+            Channel currentChannel = getChannel();
+
+            // @formatter:off
+                Channel newChannel = ChannelBuilder.create(getChannelUID(), itemType)
+                        .withDescription(description)
+                        .withKind(ChannelKind.STATE)
+                        .withLabel(getNodeId())
+                        .withType(channelTypeUID).build();
+                // @formatter:on
+
+            Thing newThing;
+            if (currentChannel != null) {
+                newThing = editThing().withoutChannel(getChannelUID()).withChannel(newChannel).build();
```

Comment:
I had the same problem in my yamaha binding. The intention is to tell the framework that a channel has changed it's structure. For me `withoutChannel(...).withChannel(...)` makes perfect sense, because there is no `updateChannel(...)`.